src/algorithms/xp.cpp

```diff
@@ -0,0 +1,724 @@
+        std::string pname;
+#ifdef VERBOSE_DEBUG
+        cerr << "counting paths" << endl;
+#endif
+        for_each_path_element([&](const std::string& path_name, const nid_t& node_id, const bool& is_rev, const std::string& cigar, const bool& is_empty, const bool& is_circular) {
+            if (path_name != pname) {
+                ++path_count;
+            }
+            pname = path_name;
+        });
+#ifdef VERBOSE_DEBUG
+        std::cerr << "graph has " << seq_length << "bp in sequence, "
+              << node_count << " nodes, "
+              << edge_count << " edges, and "
+              << path_count << " paths" << std::endl
+              << "node ids run from " << min_id << " to " << max_id << std::endl;
+#endif
+
+#ifdef VERBOSE_DEBUG
+        cerr << "storing paths" << endl;
+#endif
+        // paths
+        std::string path_names;
+
+        std::string curr_path_name;
+        std::vector<handle_t> curr_path_steps;
+        size_t curr_node_count = 0;
+        bool curr_is_circular = false; // TODO, use TP:Z:circular tag... we'll have to fish this out of the file
+        uint64_t p = 0;
+
+        auto build_accumulated_path = [&](void) {
+            // only build if we had a path to build
+#ifdef VERBOSE_DEBUG
+            if (++p % 100 == 0) std::cerr << p << " of " << path_count << " ~ " << (float)p/(float)path_count * 100 << "%" << "\r";
+#endif
+
+#ifdef debug_path_index
+            std::cerr << "Creating XPPath for path " << curr_path_name << " with visits:" << std::endl;
+        for (handle_t visiting : curr_path_steps) {
+             std::cerr << "\tHandle for " << get_id(visiting) << (get_is_reverse(visiting) ? "-" : "+") << " with value "
+                    << as_integer(visiting) << std::endl;
+        }
+#endif
+            size_t unique_member_count = 0;
+            path_names += start_marker + curr_path_name + end_marker;
+            XPPath* path = new XPPath(curr_path_name, curr_path_steps,
+                                      curr_is_circular,
+                                      *this);
+            paths.push_back(path);
+
+#ifdef debug_path_index
+            std::cerr << "paths[" << paths.size() - 1 << "] = " << curr_path_name << " @ " << path << std::endl;
+#endif
+        };
+
+        // todo ... is it circular?
+        // might make sense to scan the file for this
+        bool has_path = false;
+        for_each_path_element([&](const std::string& path_name, const nid_t& node_id, const bool& is_rev, const std::string& cigar, const bool& is_empty, const bool& is_circular) {
+            if (path_name != curr_path_name && !curr_path_name.empty()) {
+                // build the last path we've accumulated
+                build_accumulated_path();
+                curr_path_steps.clear();
+                curr_is_circular = false;
+            }
+            curr_path_name = path_name;
+            if (!is_empty) {
+                handle_t visiting = get_handle(node_id, is_rev);
```

Comment:
@ekg It seems that I need the `get_handle` function from https://github.com/vgteam/xg/blob/751f6c5552fc58055b0ca6295d132c116ed533c0/src/xg.cpp#L1629 in order to fill the path vectors. However, that implies, that I need to store `g`, `g_iv`, `g_bg_select` in XP. And these vectors must be filled and curated together with the edge vectors beforehand.
Which basically means, I have to create a full "XG" anyhow? Or do you see a way to circumvent that?